File: sdks/akta-docs/java/src/main/java/dev/openakta/aktadocs/LintEngine.java
package dev.openakta.aktadocs;

import org.commonmark.node.Document;
import org.commonmark.node.FencedCodeBlock;
import org.commonmark.node.Heading;
import org.commonmark.node.Node;
import org.commonmark.node.Paragraph;
import org.commonmark.node.Text;
import org.commonmark.parser.Parser;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.Date;
import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;

public final class LintEngine {

    private static final Pattern DOC_ID = Pattern.compile("^[a-z0-9][a-z0-9._-]*$");
    private static final Pattern ISO_DATE =
            Pattern.compile(
                    "^\\d{4}-\\d{2}-\\d{2}(?:T\\d{2}:\\d{2}:\\d{2}(?:\\.\\d+)?(?:Z|[+-]\\d{2}:\\d{2})?)?$");
    private static final Pattern QUESTION_START =
            Pattern.compile("^(how|what|when|where|why|who|which|can|should|does|is|are)\\b", Pattern.CASE_INSENSITIVE);
    private static final Pattern H23 = Pattern.compile("^(#{2,3})\\s+(.+)$");

    private static final List<String> DOC_TYPES =
            List.of(
                    "adr",
                    "business_rule",
                    "feature",
                    "guide",
                    "reference",
                    "explanation",
                    "research",
                    "meta",
                    "changelog",
                    "technical",
                    "other");

    private static final Parser PARSER = Parser.builder().build();

    private LintEngine() {}

    public static LintResult lintFiles(List<Path> paths, AktaConfig cfg, Path cwd) throws IOException {
        List<Diagnostic> diags = new ArrayList<>();
        for (Path abs : paths) {
            Path rel;
            try {
                rel = cwd.relativize(abs.toAbsolutePath().normalize());
            } catch (Exception e) {
                rel = abs.getFileName();
            }
            String relStr = rel.toString().replace('\\', '/');
            if (relStr.isEmpty()) relStr = abs.getFileName().toString();

            String raw;
            try {
                raw = Files.readString(abs);
            } catch (IOException e) {
                continue;
            }

            Frontmatter.Result fm;
            try {
                fm = Frontmatter.parse(raw);
            } catch (IOException e) {
                push(
                        diags,
                        relStr,
                        "META-001",
                        RuleSeverity.mapSeverity(RuleSeverity.effective(cfg, "META-001")),
                        1,
                        1,
                        "Invalid frontmatter YAML.");
                continue;
            }

            Map<String, Object> data = fm.data();
            String body = fm.body();

            lintMeta(relStr, raw, data, cfg, diags);
            if (!Frontmatter.hasBlock(raw)) {
                continue;
            }

            Document doc;
            try {
                doc = (Document) PARSER.parse(body);
            } catch (Exception e) {
                push(
                        diags,
                        relStr,
                        "META-001",
                        RuleSeverity.mapSeverity(RuleSeverity.effective(cfg, "META-001")),
                        1,
                        1,
                        "Markdown body failed to parse.");
                continue;
            }

            int bodyStartLine = Frontmatter.bodyStartLine(raw, body);
            int lineOffset = bodyStartLine - 1;

            lintMetaQuick(relStr, body, data, doc, cfg, diags, lineOffset);
            lintStruct008(relStr, body, data, cfg, diags, lineOffset);
            lintContent001(relStr, body, cfg, diags, lineOffset);
        }

        diags.sort(
                Comparator.comparing((Diagnostic d) -> d.file())
                        .thenComparingInt(Diagnostic::line)
                        .thenComparingInt(Diagnostic::column)
                        .thenComparing(Diagnostic::ruleId));

        int err = (int) diags.stream().filter(d -> "error".equals(d.severity())).count();
        int warn = (int) diags.stream().filter(d -> "warn".equals(d.severity())).count();
        return new LintResult(diags, new LintSummary(err, warn));
    }

    private static void push(
            List<Diagnostic> diags,
            String file,
            String rule,
            String severity,
            int line,
            int col,
            String msg) {
        if (severity == null) return;
        diags.add(new Diagnostic(file, line, col, rule, severity, msg, null, null));
    }

    private static void push(
            List<Diagnostic> diags,
            String file,
            String rule,
            String severity,
            int line,
            int col,
            String msg,
            Integer endLine,
            Integer endCol) {
        if (severity == null) return;
        diags.add(new Diagnostic(file, line, col, rule, severity, msg, endLine, endCol));
    }

    static void lintMeta(
            String file, String fileContent, Map<String, Object> data, AktaConfig cfg, List<Diagnostic> diags) {
        String r1 = RuleSeverity.effective(cfg, "META-001");
        if (!Frontmatter.hasBlock(fileContent)) {
            push(diags, file, "META-001", RuleSeverity.mapSeverity(r1), 1, 1, "Missing YAML frontmatter (expected leading --- block).");
            return;
        }

        String r2 = RuleSeverity.effective(cfg, "META-002");
        Object docId = data.get("doc_id");
        if (!(docId instanceof String) || ((String) docId).isBlank()) {
            push(diags, file, "META-002", RuleSeverity.mapSeverity(r2), 2, 1, "Frontmatter must include non-empty string `doc_id`.");
        } else if (!DOC_ID.matcher(((String) docId).strip()).matches()) {
            push(diags, file, "META-002", RuleSeverity.mapSeverity(r2), 2, 1, "doc_id must match pattern " + DOC_ID + ": got \"" + docId + "\".");
        }

        String r3 = RuleSeverity.effective(cfg, "META-003");
        Object docType = data.get("doc_type");
        if (!(docType instanceof String) || ((String) docType).isBlank()) {
            push(diags, file, "META-003", RuleSeverity.mapSeverity(r3), 2, 1, "Frontmatter must include string `doc_type`.");
        } else if (!DOC_TYPES.contains(((String) docType).strip())) {
            push(diags, file, "META-003", RuleSeverity.mapSeverity(r3), 2, 1, "Invalid doc_type \"" + docType + "\".");
        }

        String r4 = RuleSeverity.effective(cfg, "META-004");
        Object dateVal = data.get("date");
        String dateStr = null;
        if (dateVal instanceof String s && !s.isBlank()) {
            dateStr = s.strip();
        } else if (dateVal instanceof Date d) {
            dateStr = new java.text.SimpleDateFormat("yyyy-MM-dd").format(d);
        } else if (dateVal instanceof LocalDate ld) {
            dateStr = ld.toString();
        }
        if (dateStr == null) {
            push(diags, file, "META-004", RuleSeverity.mapSeverity(r4), 2, 1, "Frontmatter must include ISO8601 `date` (YYYY-MM-DD or full instant).");
        } else if (!ISO_DATE.matcher(dateStr).matches()) {
            try {
                LocalDate.parse(dateStr, DateTimeFormatter.ISO_LOCAL_DATE);
            } catch (DateTimeParseException e) {
                try {
                    java.time.OffsetDateTime.parse(dateStr, DateTimeFormatter.ISO_OFFSET_DATE_TIME);
                } catch (Exception e2) {
                    push(diags, file, "META-004", RuleSeverity.mapSeverity(r4), 2, 1, "date must be ISO8601: got \"" + dateStr + "\".");
                }
            }
        }
    }

    static void lintMetaQuick(
            String file,
            String body,
            Map<String, Object> data,
            Document doc,
            AktaConfig cfg,
            List<Diagnostic> diags,
            int lineOffset) {
        String sev = RuleSeverity.effective(cfg, "META-QUICK");
        String eff = RuleSeverity.mapSeverity(sev);
        if (eff == null) return;
        if ("changelog".equals(String.valueOf(data.get("doc_type")))) return;

        int minW = (int) RuleSeverity.numberOption(cfg, "META-QUICK", "min_words", 40);
        int maxW = (int) RuleSeverity.numberOption(cfg, "META-QUICK", "max_words", 80);

        FencedCodeBlock quick = findQuickFence(doc);
        if (quick == null) {
            push(diags, file, "META-QUICK", eff, 1 + lineOffset, 1, "After the first H1, expected a fenced code block with language tag `quick` (```quick).");
            return;
        }
        String text = quick.getLiteral() == null ? "" : quick.getLiteral().strip();
        int words = WordCount.count(text);
        int lineInBody = fenceQuickOpenLine1(body);
        int line = lineInBody + lineOffset;
        int col = 1;
        if (words < minW || words > maxW) {
            push(diags, file, "META-QUICK", eff, line, col, "Quick Answer block must be " + minW + "–" + maxW + " words; found " + words + ".", null, null);
        }
    }

    /** 1-based line number within body of the line containing ```quick */
    private static int fenceQuickOpenLine1(String body) {
        int idx = body.indexOf("```quick");
        if (idx < 0) idx = body.indexOf("``` quick");
        if (idx < 0) return 1;
        int n = 1;
        for (int i = 0; i < idx; i++) {
            if (body.charAt(i) == '\n') n++;
        }
        return n;
    }

    private static FencedCodeBlock findQuickFence(Document doc) {
        Node n = doc.getFirstChild();
        while (n != null) {
            if (n instanceof Heading h && h.getLevel() == 1) {
                n = n.getNext();
                while (n != null) {
                    if (n instanceof Heading) return null;
                    if (n instanceof FencedCodeBlock fcb) {
                        String info = trimInfo(fcb.getInfo());
                        if ("quick".equals(info)) return fcb;
                        return null;
                    }
                    if (n instanceof Paragraph p) {
                        if (paragraphBlank(p)) {
                            n = n.getNext();
                            continue;
                        }
                        return null;
                    }
                    return null;
                }
                return null;
            }
            n = n.getNext();
        }
        return null;
    }

    private static String trimInfo(String info) {
        if (info == null) return "";
        String[] parts = info.strip().split("\\s+");
        return parts.length > 0 ? parts[0] : "";
    }

    private static boolean paragraphBlank(Paragraph p) {
        Node c = p.getFirstChild();
        if (c == null) return true;
        if (c instanceof Text t) return t.getLiteral() == null || t.getLiteral().isBlank();
        return false;
    }

    static void lintStruct008(
            String file,
            String body,
            Map<String, Object> data,
            AktaConfig cfg,
            List<Diagnostic> diags,
            int lineOffset) {
        String sev = RuleSeverity.effective(cfg, "STRUCT-008");
        String eff = RuleSeverity.mapSeverity(sev);
        if (eff == null) return;
        if ("changelog".equals(String.valueOf(data.get("doc_type")))) return;

        int minW = (int) RuleSeverity.numberOption(cfg, "STRUCT-008", "min_words", 150);
        int maxW = (int) RuleSeverity.numberOption(cfg, "STRUCT-008", "max_words", 300);

        List<HeadingInfo> hs = collectH23(body);
        hs.sort(Comparator.comparingInt(HeadingInfo::line0));

        String[] lines = body.split("\\R", -1);

        for (HeadingInfo h : hs) {
            int firstContent0 = h.line0() + 1;
            HeadingInfo nxt =
                    hs.stream()
                            .filter(
                                    x ->
                                            x != h
                                                    && x.line0() > h.line0()
                                                    && x.depth() <= h.depth())
                            .findFirst()
                            .orElse(null);
            int endExclusive = nxt == null ? lines.length : nxt.line0();
            if (firstContent0 > endExclusive) {
                push(
                        diags,
                        file,
                        "STRUCT-008",
                        eff,
                        h.line0() + 1 + lineOffset,
                        1,
                        "Section has no body; cannot satisfy STRUCT-008 word range.",
                        h.line0() + 1 + lineOffset,
                        null);
                continue;
            }
            StringBuilder sb = new StringBuilder();
            for (int i = firstContent0; i < endExclusive; i++) {
                if (i > firstContent0) sb.append('\n');
                sb.append(lines[i]);
            }
            int wc = WordCount.count(sb.toString());
            int line = h.line0() + 1 + lineOffset;
            if (wc < minW || wc > maxW) {
                push(
                        diags,
                        file,
                        "STRUCT-008",
                        eff,
                        line,
                        1,
                        "H" + h.depth() + " section must be " + minW + "–" + maxW + " words; found " + wc + ".",
                        h.line0() + 1 + lineOffset,
                        null);
            }
        }
    }

    static List<HeadingInfo> collectH23(String body) {
        List<HeadingInfo> out = new ArrayList<>();
        String[] lines = body.split("\\R", -1);
        for (int i = 0; i < lines.length; i++) {
            var m = H23.matcher(lines[i]);
            if (m.matches()) {
                int depth = m.group(1).length();
                if (depth == 2 || depth == 3) {
                    out.add(new HeadingInfo(depth, i, m.group(2).strip()));
                }
            }
        }
        return out;
    }

    static void lintContent001(
            String file, String body, AktaConfig cfg, List<Diagnostic> diags, int lineOffset) {
        String sev = RuleSeverity.effective(cfg, "CONTENT-001");
        String eff = RuleSeverity.mapSeverity(sev);
        if (eff == null) return;
        double ratio = RuleSeverity.numberOption(cfg, "CONTENT-001", "min_question_ratio", 0.7);
        List<HeadingInfo> hs = collectH23(body);
        if (hs.isEmpty()) return;
        int q = 0;
        for (HeadingInfo h : hs) {
            if (isQuestion(h.text())) q++;
        }
        double actual = (double) q / hs.size();
        if (actual + 1e-9 < ratio) {
            push(
                    diags,
                    file,
                    "CONTENT-001",
                    eff,
                    1 + lineOffset,
                    1,
                    String.format(
                            "At least %.0f%% of H2/H3 headings should be questions; got %.1f%% (%d/%d).",
                            ratio * 100, actual * 100, q, hs.size()));
        }
    }

    private static boolean isQuestion(String text) {
        String s = text.strip();
        if (s.endsWith("?")) return true;
        return QUESTION_START.matcher(s).find();
    }
}

File: sdks/akta-docs/csharp/LintEngine.cs
using System.Globalization;
using System.Text.RegularExpressions;
using Markdig;
using Markdig.Syntax;
using Markdig.Syntax.Inlines;

namespace OpenAkta.AktaDocs;

public static class LintEngine
{
    private static readonly Regex DocId = new(@"^[a-z0-9][a-z0-9._-]*$", RegexOptions.Compiled);
    private static readonly Regex IsoDate = new(
        @"^\d{4}-\d{2}-\d{2}(?:T\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:Z|[+-]\d{2}:\d{2})?)?$",
        RegexOptions.Compiled);

    private static readonly Regex QuestionStart = new(
        @"^(how|what|when|where|why|who|which|can|should|does|is|are)\b",
        RegexOptions.IgnoreCase | RegexOptions.Compiled);

    private static readonly Regex H23 = new(@"^(#{2,3})\s+(.+)$", RegexOptions.Compiled);

    private static readonly string[] DocTypes =
    {
        "adr", "business_rule", "feature", "guide", "reference", "explanation", "research", "meta",
        "changelog", "technical", "other"
    };

    private static readonly MarkdownPipeline Pipeline = new MarkdownPipelineBuilder()
        .UseAdvancedExtensions()
        .Build();

    public static LintResult LintFiles(IReadOnlyList<string> paths, AktaConfig cfg, string cwd)
    {
        var diags = new List<Diagnostic>();
        foreach (var p in paths)
        {
            var abs = Path.GetFullPath(p);
            string rel;
            try
            {
                rel = Path.GetRelativePath(cwd, abs);
            }
            catch
            {
                rel = Path.GetFileName(abs);
            }

            rel = rel.Replace('\\', '/');
            if (string.IsNullOrEmpty(rel)) rel = Path.GetFileName(abs);

            string raw;
            try
            {
                raw = File.ReadAllText(abs);
            }
            catch
            {
                continue;
            }

            Dictionary<string, object?> data;
            string body;
            try
            {
                (data, body) = Frontmatter.Parse(raw);
            }
            catch (InvalidDataException)
            {
                Push(diags, rel, "META-001", RuleSeverity.MapSeverity(RuleSeverity.Effective(cfg, "META-001")),
                    1, 1, "Invalid frontmatter YAML.");
                continue;
            }

            LintMeta(rel, raw, data, cfg, diags);
            if (!Frontmatter.HasBlock(raw))
                continue;

            MarkdownDocument? doc;
            try
            {
                doc = Markdown.Parse(body, Pipeline) as MarkdownDocument;
            }
            catch
            {
                Push(diags, rel, "META-001", RuleSeverity.MapSeverity(RuleSeverity.Effective(cfg, "META-001")),
                    1, 1, "Markdown body failed to parse.");
                continue;
            }

            if (doc == null) continue;

            var bodyStartLine = Frontmatter.BodyStartLine(raw, body);
            var lineOffset = bodyStartLine - 1;

            LintMetaQuick(rel, body, data, doc, cfg, diags, lineOffset);
            LintStruct008(rel, body, data, cfg, diags, lineOffset);
            LintContent001(rel, body, cfg, diags, lineOffset);
        }

        diags.Sort(CompareDiag);
        var err = diags.Count(d => d.Severity == "error");
        var warn = diags.Count(d => d.Severity == "warn");
        return new LintResult(diags, new LintSummary(err, warn));
    }

    private static int CompareDiag(Diagnostic a, Diagnostic b)
    {
        var c = string.Compare(a.File, b.File, StringComparison.Ordinal);
        if (c != 0) return c;
        if (a.Line != b.Line) return a.Line.CompareTo(b.Line);
        if (a.Column != b.Column) return a.Column.CompareTo(b.Column);
        return string.Compare(a.RuleId, b.RuleId, StringComparison.Ordinal);
    }

    private static void Push(
        List<Diagnostic> diags, string file, string rule, string? severity,
        int line, int col, string msg, int? endLine = null, int? endCol = null)
    {
        if (severity == null) return;
        diags.Add(new Diagnostic(file, line, col, rule, severity, msg, endLine, endCol));
    }

    private static void LintMeta(
        string file, string fileContent, IReadOnlyDictionary<string, object?> data, AktaConfig cfg,
        List<Diagnostic> diags)
    {
        var r1 = RuleSeverity.Effective(cfg, "META-001");
        if (!Frontmatter.HasBlock(fileContent))
        {
            Push(diags, file, "META-001", RuleSeverity.MapSeverity(r1), 1, 1,
                "Missing YAML frontmatter (expected leading --- block).");
            return;
        }

        var r2 = RuleSeverity.Effective(cfg, "META-002");
        data.TryGetValue("doc_id", out var docIdObj);
        var docId = docIdObj?.ToString();
        if (string.IsNullOrWhiteSpace(docId))
        {
            Push(diags, file, "META-002", RuleSeverity.MapSeverity(r2), 2, 1,
                "Frontmatter must include non-empty string `doc_id`.");
        }
        else if (!DocId.IsMatch(docId.Trim()))
        {
            Push(diags, file, "META-002", RuleSeverity.MapSeverity(r2), 2, 1,
                $"doc_id must match pattern {DocId}: got \"{docId}\".");
        }

        var r3 = RuleSeverity.Effective(cfg, "META-003");
        data.TryGetValue("doc_type", out var docTypeObj);
        var docType = docTypeObj?.ToString();
        if (string.IsNullOrWhiteSpace(docType))
        {
            Push(diags, file, "META-003", RuleSeverity.MapSeverity(r3), 2, 1,
                $"Frontmatter must include string `doc_type` (one of: {string.Join(", ", DocTypes)}).");
        }
        else if (!DocTypes.Contains(docType.Trim(), StringComparer.Ordinal))
        {
            Push(diags, file, "META-003", RuleSeverity.MapSeverity(r3), 2, 1, $"Invalid doc_type \"{docType}\".");
        }

        var r4 = RuleSeverity.Effective(cfg, "META-004");
        data.TryGetValue("date", out var dateVal);
        var dateStr = CoerceDateString(dateVal);
        if (string.IsNullOrEmpty(dateStr))
        {
            Push(diags, file, "META-004", RuleSeverity.MapSeverity(r4), 2, 1,
                "Frontmatter must include ISO8601 `date` (YYYY-MM-DD or full instant).");
        }
        else if (!IsoDate.IsMatch(dateStr))
        {
            if (!DateTime.TryParse(dateStr, CultureInfo.InvariantCulture,
                    DateTimeStyles.RoundtripKind, out _)
                && !DateTimeOffset.TryParse(dateStr, CultureInfo.InvariantCulture,
                    DateTimeStyles.RoundtripKind, out _))
            {
                Push(diags, file, "META-004", RuleSeverity.MapSeverity(r4), 2, 1,
                    $"date must be ISO8601: got \"{dateStr}\".");
            }
        }
    }

    private static string? CoerceDateString(object? v)
    {
        if (v == null) return null;
        return v switch
        {
            string s when !string.IsNullOrWhiteSpace(s) => s.Trim(),
            DateTime dt => dt.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture),
            DateTimeOffset dto => dto.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture),
            _ => v.ToString()?.Trim()
        };
    }

    private static void LintMetaQuick(
        string file, string body, IReadOnlyDictionary<string, object?> data, MarkdownDocument doc,
        AktaConfig cfg, List<Diagnostic> diags, int lineOffset)
    {
        var sev = RuleSeverity.Effective(cfg, "META-QUICK");
        var eff = RuleSeverity.MapSeverity(sev);
        if (eff == null) return;
        if (data.TryGetValue("doc_type", out var dt) && string.Equals(dt?.ToString(), "changelog", StringComparison.Ordinal))
            return;

        var minW = (int)RuleSeverity.NumberOption(cfg, "META-QUICK", "min_words", 40);
        var maxW = (int)RuleSeverity.NumberOption(cfg, "META-QUICK", "max_words", 80);

        var quick = FindQuickFence(doc);
        if (quick == null)
        {
            Push(diags, file, "META-QUICK", eff, 1 + lineOffset, 1,
                "After the first H1, expected a fenced code block with language tag `quick` (```quick).");
            return;
        }

        var text = FenceLiteral(quick).Trim();
        var words = WordCount.Count(text);
        var line = FenceQuickOpenLine1(body) + lineOffset;
        const int col = 1;
        if (words < minW || words > maxW)
        {
            Push(diags, file, "META-QUICK", eff, line, col,
                $"Quick Answer block must be {minW}–{maxW} words; found {words}.");
        }
    }

    private static int FenceQuickOpenLine1(string body)
    {
        var idx = body.IndexOf("```quick", StringComparison.Ordinal);
        if (idx < 0) idx = body.IndexOf("``` quick", StringComparison.Ordinal);
        if (idx < 0) return 1;
        var n = 1;
        for (var i = 0; i < idx; i++)
        {
            if (body[i] == '\n') n++;
        }

        return n;
    }

    private static string FenceLiteral(FencedCodeBlock f)
    {
        var sb = new System.Text.StringBuilder();
        foreach (var line in f.Lines)
            sb.Append(line);
        return sb.ToString();
    }

    private static FencedCodeBlock? FindQuickFence(MarkdownDocument doc)
    {
        var foundH1 = false;
        foreach (var block in doc)
        {
            if (!foundH1)
            {
                if (block is HeadingBlock { Level: 1 })
                    foundH1 = true;
                continue;
            }

            if (block is HeadingBlock) return null;
            if (block is FencedCodeBlock fcb)
            {
                var info = TrimInfo(fcb.Info);
                return info == "quick" ? fcb : null;
            }

            if (block is ParagraphBlock pb)
            {
                if (ParagraphBlank(pb)) continue;
                return null;
            }

            return null;
        }

        return null;
    }

    private static string TrimInfo(string? info)
    {
        if (string.IsNullOrWhiteSpace(info)) return "";
        var parts = info.Trim().Split(' ', StringSplitOptions.RemoveEmptyEntries);
        return parts.Length > 0 ? parts[0] : "";
    }

    private static bool ParagraphBlank(ParagraphBlock p)
    {
        if (p.Inline == null) return true;
        for (var il = p.Inline.FirstChild; il != null; il = il.NextSibling)
        {
            switch (il)
            {
                case LiteralInline lit:
                {
                    foreach (var ch in lit.Content.AsSpan())
                    {
                        if (!char.IsWhiteSpace(ch)) return false;
                    }

                    break;
                }
                case LineBreakInline:
                    break;
                default:
                    return false;
            }
        }

        return true;
    }

    private static void LintStruct008(
        string file, string body, IReadOnlyDictionary<string, object?> data, AktaConfig cfg,
        List<Diagnostic> diags, int lineOffset)
    {
        var sev = RuleSeverity.Effective(cfg, "STRUCT-008");
        var eff = RuleSeverity.MapSeverity(sev);
        if (eff == null) return;
        if (data.TryGetValue("doc_type", out var dt) && string.Equals(dt?.ToString(), "changelog", StringComparison.Ordinal))
            return;

        var minW = (int)RuleSeverity.NumberOption(cfg, "STRUCT-008", "min_words", 150);
        var maxW = (int)RuleSeverity.NumberOption(cfg, "STRUCT-008", "max_words", 300);

        var hs = CollectH23(body);
        hs.Sort((a, b) => a.Line0.CompareTo(b.Line0));

        var lines = body.Split(new[] { "\r\n", "\n", "\r" }, StringSplitOptions.None);

        foreach (var h in hs)
        {
            var firstContent0 = h.Line0 + 1;
            HeadingInfo? nxt = null;
            foreach (var x in hs)
            {
                if (x.Line0 > h.Line0 && x.Depth <= h.Depth)
                {
                    nxt = x;
                    break;
                }
            }

            var endExclusive = nxt == null ? lines.Length : nxt.Value.Line0;
            if (firstContent0 > endExclusive)
            {
                Push(diags, file, "STRUCT-008", eff, h.Line0 + 1 + lineOffset, 1,
                    "Section has no body; cannot satisfy STRUCT-008 word range.",
                    h.Line0 + 1 + lineOffset, null);
                continue;
            }

            var sb = new System.Text.StringBuilder();
            for (var i = firstContent0; i < endExclusive; i++)
            {
                if (i > firstContent0) sb.Append('\n');
                sb.Append(lines[i]);
            }

            var wc = WordCount.Count(sb.ToString());
            var line = h.Line0 + 1 + lineOffset;
            if (wc < minW || wc > maxW)
            {
                Push(diags, file, "STRUCT-008", eff, line, 1,
                    $"H{h.Depth} section must be {minW}–{maxW} words; found {wc}.",
                    h.Line0 + 1 + lineOffset, null);
            }
        }
    }

    private static List<HeadingInfo> CollectH23(string body)
    {
        var lines = body.Split(new[] { "\r\n", "\n", "\r" }, StringSplitOptions.None);
        var outList = new List<HeadingInfo>();
        for (var i = 0; i < lines.Length; i++)
        {
            var m = H23.Match(lines[i]);
            if (!m.Success) continue;
            var depth = m.Groups[1].Value.Length;
            if (depth is 2 or 3)
                outList.Add(new HeadingInfo(depth, i, m.Groups[2].Value.Trim()));
        }

        return outList;
    }

    private static void LintContent001(
        string file, string body, AktaConfig cfg, List<Diagnostic> diags, int lineOffset)
    {
        var sev = RuleSeverity.Effective(cfg, "CONTENT-001");
        var eff = RuleSeverity.MapSeverity(sev);
        if (eff == null) return;
        var ratio = RuleSeverity.NumberOption(cfg, "CONTENT-001", "min_question_ratio", 0.7);
        var hs = CollectH23(body);
        if (hs.Count == 0) return;
        var q = hs.Count(IsQuestion);
        var actual = (double)q / hs.Count;
        if (actual + 1e-9 < ratio)
        {
            Push(diags, file, "CONTENT-001", eff, 1 + lineOffset, 1,
                $"At least {ratio * 100:0}% of H2/H3 headings should be questions; got {actual * 100:0.0}% ({q}/{hs.Count}).");
        }
    }

    private static bool IsQuestion(HeadingInfo h) => IsQuestionText(h.Text);

    private static bool IsQuestionText(string text)
    {
        var s = text.Trim();
        if (s.EndsWith('?')) return true;
        return QuestionStart.IsMatch(s);
    }
}
